internal/engine/compiler/impl_amd64.go

```diff
@@ -3598,21 +3598,49 @@ func (c *amd64Compiler) compileLoadDataInstanceAddress(dataIndex uint32, dst asm
 	)
 }
 
-// compileMemoryCopy implements compiler.compileMemoryCopy for the amd64 architecture.
-func (c *amd64Compiler) compileMemoryCopy() error {
-	return c.compileCopyImpl(false, 0, 0)
-}
+func (c *amd64Compiler) compileCopyLoopImpl(destinationOffset, sourceOffset, copySize *runtimeValueLocation, backwards bool, bwOffset uint8) {
+	// skip if nothing to copy
+	c.assembler.CompileRegisterToConst(amd64.CMPQ, copySize.register, 0)
+	emptyEightGroupsJump := c.assembler.CompileJump(amd64.JEQ)
 
-// compileCopyImpl implements compileTableCopy and compileMemoryCopy.
-//
-// TODO: the compiled code in this function should be reused and compile at once as
-// the code is independent of any module.
-func (c *amd64Compiler) compileCopyImpl(isTable bool, srcTableIndex, dstTableIndex uint32) error {
-	outOfBoundsErrorStatus := nativeCallStatusCodeMemoryOutOfBounds
-	if isTable {
-		outOfBoundsErrorStatus = nativeCallStatusCodeInvalidTableAccess
+	rvalues := []*runtimeValueLocation{destinationOffset, sourceOffset, copySize}
+
+	// prepare registers for swaps. There will never be more than 3 XCHGs in total
+	swaps := c.compilePreventCrossedTargetRegisters(rvalues, []asm.Register{amd64.RegDI, amd64.RegSI, amd64.RegCX})
+
+	// prepare registers for REP MOVSQ: copy from rsi to rdi, rcx times
+	c.compileSwapRegisters(destinationOffset.register, amd64.RegDI)
+	c.compileSwapRegisters(sourceOffset.register, amd64.RegSI)
+	c.compileSwapRegisters(copySize.register, amd64.RegCX)
+
+	// point on first quadword to copy
+	if backwards {
+		c.assembler.CompileConstToRegister(amd64.ADDQ, -int64(bwOffset), amd64.RegDI)
+		c.assembler.CompileConstToRegister(amd64.ADDQ, -int64(bwOffset), amd64.RegSI)
+		// rep direction backwards
+		c.assembler.CompileStandAlone(amd64.STD)
 	}
 
+	c.assembler.CompileStandAlone(amd64.REPMOVSQ)
+
+	if backwards {
+		// reset direction
+		c.assembler.CompileStandAlone(amd64.CLD)
+	}
+
+	// restore registers
+	c.compileSwapRegisters(destinationOffset.register, amd64.RegDI)
+	c.compileSwapRegisters(sourceOffset.register, amd64.RegSI)
+	c.compileSwapRegisters(copySize.register, amd64.RegCX)
+
+	c.compileRestorePreventedCrossing(rvalues, swaps)
+
+	c.assembler.SetJumpTargetOnNext(emptyEightGroupsJump)
+	c.assembler.CompileStandAlone(amd64.NOP)
+}
+
+// compileMemoryCopy implements compiler.compileMemoryCopy for the amd64 architecture.
+func (c *amd64Compiler) compileMemoryCopy() error {
```

Comment:
Could you add the detailed doc about the logic in this function so that others can understand what kind of native code will be generated here without looking at the function body? e.g. Pseudo assembly etc.